File: StockPro-Demo-Backend/microservices/purchase-service/src/main/java/com/stockpro/purchase/service/PurchaseService.java
package com.stockpro.purchase.service;

import com.stockpro.purchase.dto.PartialReceiptItemDTO;
import com.stockpro.purchase.dto.PurchaseOrderRequestDTO;
import com.stockpro.purchase.dto.PurchaseOrderResponseDTO;
import com.stockpro.purchase.entity.POStatus;

import java.time.LocalDate;
import java.util.List;

public interface PurchaseService {
    List<PurchaseOrderResponseDTO> getAllPOs();
    PurchaseOrderResponseDTO createPO(PurchaseOrderRequestDTO requestDTO);
    PurchaseOrderResponseDTO getPOById(Integer poId);
    List<PurchaseOrderResponseDTO> getPOsBySupplier(Integer supplierId);
    List<PurchaseOrderResponseDTO> getPOsByWarehouse(Integer warehouseId);
    List<PurchaseOrderResponseDTO> getPOsByStatus(POStatus status);
    List<PurchaseOrderResponseDTO> getPOsByDateRange(LocalDate startDate, LocalDate endDate);
    PurchaseOrderResponseDTO updatePO(Integer poId, PurchaseOrderRequestDTO requestDTO);
    PurchaseOrderResponseDTO approvePO(Integer poId);
    PurchaseOrderResponseDTO cancelPO(Integer poId);
    PurchaseOrderResponseDTO receiveGoods(Integer poId);                                      // Full receipt
    PurchaseOrderResponseDTO receiveGoodsPartially(Integer poId, List<PartialReceiptItemDTO> items); // Partial receipt
}

File: stockpro-frontend/src/services/purchaseService.js
import { purchaseApi } from './api';

export const purchaseService = {
    // Get all POs
    getAllPOs: async () => {
        const response = await purchaseApi.get('/purchase-orders');
        return response.data;
    },

    // Get PO by ID
    getPOById: async (id) => {
        const response = await purchaseApi.get(`/purchase-orders/${id}`);
        return response.data;
    },

    // Create a new PO
    createPO: async (poData) => {
        const response = await purchaseApi.post('/purchase-orders', poData);
        return response.data;
    },

    // Get POs by status
    getPOsByStatus: async (status) => {
        const response = await purchaseApi.get(`/purchase-orders/status/${status}`);
        return response.data;
    },

    // Approve a PO
    approvePO: async (id) => {
        const response = await purchaseApi.put(`/purchase-orders/${id}/approve`);
        return response.data;
    },

    // Cancel a PO
    cancelPO: async (id) => {
        const response = await purchaseApi.put(`/purchase-orders/${id}/cancel`);
        return response.data;
    },

    // Full Receipt
    receiveGoods: async (id) => {
        const response = await purchaseApi.post(`/purchase-orders/${id}/receive`);
        return response.data;
    },

    // Partial Receipt
    // items should be array of { poItemId, receivedQuantity }
    receiveGoodsPartially: async (id, items) => {
        const response = await purchaseApi.post(`/purchase-orders/${id}/receive/partial`, items);
        return response.data;
    }
};
